### ISSUE TITLE
MunkiReport v4 and MySQL using SSL

### ISSUE THREAD

**Issue by jelockwood** [munkireport/munkireport-php#1217]
I am currently running MunkiReport-PHP v3.2 and have it connected to a MySQL server using username/password authentication _and_ requiring an SSL encrypted connection. This is working fine. Attempts to connect without a valid SSL certificate connection are rejected and hence fail as desired with v3.2.

The snipit from the config.php for this is as follows.

```
        $conf['pdo_dsn'] = 'mysql:host=mysql.domain.com;dbname=munkireport;port=3306';
        $conf['pdo_user'] = 'munkireport';
        $conf['pdo_pass'] = 'munkireport';
        $conf['pdo_opts'] = array(
                        PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8',
                        PDO::MYSQL_ATTR_SSL_KEY => '/etc/mysql/ssl/munki.key',
                        PDO::MYSQL_ATTR_SSL_CERT => '/etc/mysql/ssl/munki.crt',
                        PDO::MYSQL_ATTR_SSL_CA => '/etc/mysql/ssl/CA.crt'
                                );
```

My problem is that MunkiReport-PHP uses a .env config file and I cannot see any information in either the example .env or the wiki page at 
https://github.com/munkireport/munkireport-php/wiki/MunkiReport-Database#MySQL on how to set the equivalent SSL options.

(By the way there are still some instances in the MunkiReport wiki pages including the above where the use of the old style $conf variables is still referenced. At least on the above MySQL wiki page this seems incorrect.)

Since I have not been able to define the same SSL options v4 is not allowed to connect to the MySQL server and database.

**Comment by bochoven** [comment]
We should be able to add that to MR

**Comment by bochoven** [comment]
Could you check if the changes in https://github.com/munkireport/munkireport-php/commit/182ed251cf068ffcac9b23c64ba866a1d89c4fa0 address your issue?

**Comment by jelockwood** [comment]
Ok, the original error was as follows -

```
Fatal error

Connection failed: SQLSTATE[HY000] [1045] Access denied for user 'munkireport'@'mysql.domain.com' (using password: YES)
Try again
```

I now have the (sanitised) equivalent of the following in my .env

```
# --------------------------------------------
# OPTIONAL: SSL DATABASE SETTINGS
# --------------------------------------------
DB_SSL='true'
DB_SSL_KEY_PATH='/etc/mysql/ssl/munki.domain.com.key'
DB_SSL_CERT_PATH='/etc/mysql/ssl/munki.domain.com.crt'
DB_SSL_CA_PATH='/etc/mysql/ssl/CA.crt'
DB_SSL_CIPHER='DHE-RSA-AES256-SHA:AES128-SHA'
```

I then noticed you actually chose different names so changed the above to

```
# --------------------------------------------
# OPTIONAL: SSL DATABASE SETTINGS
# --------------------------------------------
# DB_SSL='true'
CONNECTION_SSL_KEY='/etc/mysql/ssl/munki.domain.com.key'
CONNECTION_SSL_CERT='/etc/mysql/ssl/munki.domain.com.crt'
CONNECTION_SSL_CA_PATH='/etc/mysql/ssl/CA.crt'
CONNECTION_SSL_CIPHER='DHE-RSA-AES256-SHA:AES128-SHA'
```

I do not know what the intent of

CONNECTION_SSL_CA

is for. This on the face of it would seem to duplicate the function of CONNECTION_SSL_CA_PATH

I have manually edited the following files -

app/config/db.php
app/helpers/site_helper.php

To add the additional lines.

I am now getting a different error as follows -

```
Warning: constant(): Couldn't find constant PDO::MYSQL_ATTR_SSL_key in /var/www/munkireport-php/app/helpers/site_helper.php on line 94

Warning: constant(): Couldn't find constant PDO::MYSQL_ATTR_SSL_cert in /var/www/munkireport-php/app/helpers/site_helper.php on line 94

Warning: constant(): Couldn't find constant PDO::MYSQL_ATTR_SSL_capath in /var/www/munkireport-php/app/helpers/site_helper.php on line 94

Warning: constant(): Couldn't find constant PDO::MYSQL_ATTR_SSL_cipher in /var/www/munkireport-php/app/helpers/site_helper.php on line 94
Fatal error

Connection failed: SQLSTATE[HY000] [1045] Access denied for user 'munkireport'@'mysql.domain.com' (using password: YES)
Try again
```

**Comment by bochoven** [comment]
I `updated app/helpers/site_helper.php` to fix the `constant` errors.

**Comment by bochoven** [comment]
The difference between `CONNECTION_SSL_CA ` and `CONNECTION_SSL_CA_PATH` is 

```
PDO::MYSQL_ATTR_SSL_CA (integer)
The file path to the SSL certificate authority.

PDO::MYSQL_ATTR_SSL_CAPATH (integer)
The file path to the directory that contains the trusted SSL CA certificates, which are stored in PEM format.
```

as explained here: http://php.net/manual/en/ref.pdo-mysql.php

**Comment by bochoven** [comment]
Thanks for the feedback, I don't have an SSL equipped MySQL instance to test with.

By the way, the PDO constants are, in fact, integers: 

```
php -r 'echo PDO::MYSQL_ATTR_SSL_CA;'
1009
```

**Comment by jelockwood** [comment]
Ok, I logged in as an admin level user to munkireport, I then went to the Admin menu and selected 'Database'. I then told it to do an update.

The only entry I am getting is -

```
Upgrade Log
--
Started update: Thu Jan 24 2019 15:35:10 GMT+0000 (Greenwich Mean Time)
```

Trying the command line option of 

php database/migrate.php

results in -

`SQLSTATE[HY000] [1045] Access denied for user 'munkireport'@'mysql.domain.com' (using password: YES) (SQL: select * from information_schema.tables where table_schema = munkireport and table_name = migrations)`

Obviously the web option of using the Admin menu should be using the now added SSL options but the command line error suggests it is not seeing the SSL options.

Any suggestions on how to get it successfully upgrade the database?

I know that now with the SSL options munkireport is getting some access to the database because I can see some entries. I however need to do the update because it is part of upgrading and because it has not yet been done some of the tables e.g. Security in the Listing menu gives an error.

**Comment by jelockwood** [comment]
I managed to find and remember how to edit the login access in mysql to add/remove SSL as a requirement in the main mysql user table. I then temporarily disabled SSL as a requirement and was able to successfully do the php database/migrate.php command.

However for the future it would be desirable to 'fix' things so this command also can use defined SSL credentials.

**Comment by jelockwood** [comment]
@bochoven 
I had to today temporarily turn off MySQL using SSL so I could install new SSL certificates on all my servers. To do this on a different app I changed the equivalent of your -

CONNECTION_SSL_ENABLED='true'

To 'false' and that was all I needed to do. However when I cam to do this for MunkiReport-PHP setting it to 'false' in your .env did not disable SSL. I had instead to comment out all the new CONNECTION_SSL lines.

It therefore seems your code is not properly checking the status of the CONNECTION_SSL_ENABLED='true' (or 'false') variable.
